### PR TITLE
Service history module

### PR DESCRIPTION
**Pull request by MonishaSivanathan**
- added controller for service_history, custom_message, lab_result
- added few fields in encounter, patient
- added DTO for master table

### CODE REVIEW

File: utils/exception.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import logging
+import traceback
+from fastapi import Request, status
+from fastapi import FastAPI, Request, HTTPException
+from fastapi.responses import JSONResponse
+from fastapi.exceptions import RequestValidationError
+
+app = FastAPI()
```
Comment (tamsud) [inline utils/exception.py:8]: why fastapi initialization is here?

File: utils/exception.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        logger.error(f"HTTP Exception: {exc.detail}")
+        logger.error(traceback.format_exc())
+        error_response_data = {
+            "error": "HTTP Exception",
+            "details": exc.detail,
+        }
+        return JSONResponse(content=error_response_data, status_code=exc.status_code)
+
+    @staticmethod
+    async def validation_exception_handler(request: Request, exc: RequestValidationError):
```
Comment (tamsud) [inline utils/exception.py:24]: is that exception added anywhwere?

**Review by tamsud — CHANGES_REQUESTED** [review]

File: constants.py [hunk 3]
```diff
@@ -11,3 +11,29 @@
 STATUS = "active"
 ENCOUNTER_TYPE_SYSTEM = "http://terminology.hl7.org/CodeSystem/v3-ActCode"
 ENCOUNTER_TYPE_CODE = "IMP"
+CURRENT_CONDITION = "Current"
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply constants.py]: added common constants

File: constants.py [hunk 4]
```diff
@@ -11,3 +11,29 @@
...
+OTHER_MEDICATION = "Other"
+RACE_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race"
+ETHNICITY_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity"
+HEIGHT_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-height"
+WEIGHT_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-weight"
+PATIENT_EXTENTION_URL = "ombCategory"
+EXTENTION_CODE = "2106-3"
+EXTENTION_SYSTEM = "urn:oid:2.16.840.1.113883.6.238"
+RACE_TEXT = "text"
+RACE_SYSTEM = "urn:oid:2.16.840.1.113883.6.238"
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply constants.py]: changed

File: controller/condition_allergy_controller.py [hunk 5]
```diff
@@ -104,27 +179,46 @@ def get_condition_by_patient_id(patient_id: str):
...
+                        ]
+                    ),
                     subject=Reference(reference=f"{PATIENT_REFERENCE}/{patient_id}"),
                 )
                 current_condition.save()
             else:
                 current_condition = None
 
-            if con.additional_condition and any(concept.display for concept in con.additional_condition):
+            if con.additional_condition and any(
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/condition_allergy_controller.py]: changed the line

File: constants.py [hunk 6]
```diff
@@ -11,3 +11,22 @@
 STATUS = "active"
 ENCOUNTER_TYPE_SYSTEM = "http://terminology.hl7.org/CodeSystem/v3-ActCode"
 ENCOUNTER_TYPE_CODE = "IMP"
+RACE_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race"
+ETHNICITY_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity"
+HEIGHT_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-height"
+WEIGHT_URL = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-weight"
+TEXT = "text"
+SYSTEM = "urn:oid:2.16.840.1.113883.6.238"
+CODE = "2106-3"
```
Comment (tamsud) [inline constants.py:20]: why code still there?

File: controller/custom_message_controller.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    def get_custom_message():
+        cwd = os.getcwd()
+        relative_path = os.path.join(cwd, MESSAGE_PATH)
+        logger.info(f"Reading message from file: {relative_path}")
+        try:
+            with open(relative_path, "r") as file:
+                logger.info(f"Reading message from file: {relative_path}")
+                data = json.load(file)
+                logger.info(f"Message read from file: {data}")
+                message = data.get("message", "No message found")
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/custom_message_controller.py]: directly sent the responce

File: controller/patient_controller.py [hunk 8]
```diff
@@ -61,6 +77,38 @@ def create_patient(pat: PatientModel):
...
+                                code=RACE_CODE,
+                                display=pat.race,
+                            ),
+                        )
+                        ],
+                    ),Extension(
+                        url=ETHNICITY_URL,
+                        extension=[
+                            Extension(
+                                url=RACE_TEXT,
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/patient_controller.py]: changed into common text

File: controller/patient_controller.py [hunk 9]
```diff
@@ -61,6 +77,38 @@ def create_patient(pat: PatientModel):
...
                 ],
+                extension=[Extension(
+                    url=RACE_URL,
+                    extension=[
+                        Extension(
+                            url=RACE_TEXT,
+                            valueString = pat.race,
+                            valueCoding=Coding(
+                                system=RACE_SYSTEM,
+                                code=RACE_CODE,
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/patient_controller.py]: nope, i have removed the code

File: controller/patient_controller.py [hunk 10]
```diff
@@ -198,73 +275,115 @@ def delete_patient_by_id(patient_id: str):
...
+            extracted_data["middleName"] = None
+            extracted_data["lastName"] = None
 
         # Extract gender
-        extracted_data['gender'] = patient.gender
+        extracted_data["gender"] = patient.gender
 
         # Extract address components
         if patient.address:
-            extracted_data['address'] = patient.address[0].text
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/patient_controller.py:234]: i have used extention for this

File: controller/service_history_controller.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            service_history = ObservationClient.get_lab_result_by_patient_id(patient_id)
+            immuzation = HL7ImmunizationClient.get_immunizations_by_patient_id(
+                patient_id
+            )
+            if all_service == True:
+                return ServiceHistoryClient.create_final_dictionary(
+                    service_history, immuzation
+                )
+        
+            if lab_result == True:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/service_history_controller.py]: changed into proper tamil.

File: controller/service_history_controller.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                    {
+                        "Id": resource_id,
+                        "Category": "Lab Result",
+                        "ServiceFor": service_display,
+                        "DateOfService": occurrence_date,
+                    }
+                )
+        return service_history_entries
+
+    def create_final_dictionary(service_history, immunization):
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/service_history_controller.py]: i have changed the common name

File: utils/exception.py (hunk 1 above)
Comment (MonishaSivanathan) [reply utils/exception.py:8]: remove the file

File: utils/exception.py (hunk 2 above)
Comment (MonishaSivanathan) [reply utils/exception.py:24]: nope, added for my personal use

**Review by tamsud — CHANGES_REQUESTED** [review]

File: constants.py (hunk 6 above)
Comment (MonishaSivanathan) [reply constants.py:20]: done

Comment (tamsud) [reply constants.py:20]: still i can see the code variable

File: utils/exception.py (hunk 2 above)
Comment (tamsud) [reply utils/exception.py:24]: file is still there

**Review by tamsud — CHANGES_REQUESTED** [review]

File: utils/exception.py (hunk 1 above)
Comment (tamsud) [reply utils/exception.py:8]: not removed yet

**Review by tamsud — CHANGES_REQUESTED** [review]

**Review by tamsud — APPROVED** [review]

**Pull request merged**